doc/namespace.md

```diff
@@ -0,0 +1,417 @@
+
+```ruby
+# foo.rb
+class C; end
+
+# main.rb
+ns = Namespace.new
+ns.require('foo')
+
+p ns::C # "#<Namespace:0x00000001083a5660>::C"
+```
+
+And currently, if a namespace is assigned to a constant `NS1`, the classpath output will be `NS1::C`. But the namespace object can be brought to another namespace and the constant `NS1` in the namespace is something different. So the constant-based classpath for namespace is not safe basically.
+
+So we should find a better format to show namespaces. Options are:
+
+* `NS1::C` (only when this namespace is created and assigned to NS1 in the current namespace)
+* `#<Namespace:user:1083a5660>::C` (with namespace type and without preceding 0)
+* or something else
+
+#### Namespace#eval
+
+Testing namespace features needs to create files to be loaded in namespaces. It's not easy nor casual.
+
+If `Namespace` class has an instance method `#eval` to evaluate code in the namespace, it can be helpful.
+
+#### More builtin methods written in Ruby
+
+If namespace is enabled by default, builtin methods can be written in Ruby because it can't be overridden by users' monkey patches. Builtin Ruby methods can be JIT-ed, and it could bring performance reward.
+
+#### Monkey patching methods called by builtin methods
+
+Builtin methods sometimes call other builtin methods. For example, `Hash#map` calls `Hash#each` to retrieve entries to be mapped. Without namespace, Ruby users can overwrite `Hash#each` and expect the behavior change of `Hash#map` as a result.
+
+But with namespaces, `Hash#map` runs in the root namespace. Ruby users can define `Hash#each` only in user namespaces, so users cannot change `Hash#map`'s behavior in this case. To achieve it, users should override both`Hash#map` and `Hash#each` (or only `Hash#map`).
+
+It is a breaking change.
+
+It's an option to change the behavior of methods in the root namespace to refer to definitions in user namespaces. But if we do so, that means we can't proceed with "More builtin methods written in Ruby".
+
+#### Context of \$LOAD\_PATH and \$LOADED\_FEATURES
+
+Global variables `$LOAD_PATH` and `$LOADED_FEATURES` control `require` method behaviors. So those namespaces are determined by the loading namespace instead of the current namespace.
+
+This could potentially conflict with the user's expectations. We should find the solution.
+
+#### Expose top level methods as a method of the namespace object
+
+Currently, top level methods in namespaces are not accessible from outside of the namespace. But there might be a use case to call other namespace's top level methods.
+
+#### Split root and builtin namespace
+
+NOTE: "builtin" namespace is a different one from the "builtin" namespace in the current implementation
+
+Currently, the single "root" namespace is the source of classext CoW. And also, the "root" namespace can load additional files after starting main script evaluation by calling methods which contain lines like `require "openssl"`.
```

Comment:
The root namespace, assuming it contains only the core library/builtin classes should have no method doing any `require`, isn't it?